Support for custom HTTP headers

qharnay opened the pull request:
Hello, and thanks for your work on EventSourcing from backend with Promises ! 

I'd like to use this lib. to set up a backend subscriber, but there was a lack of support with custom headers, preventing me to subscribe to my SSE server (I have to provide Authorization Bearers header into HTTP requests).

This PR enables custom headers to be sent into HTTP headers of the EventSource's request.

Tests have been updated accordingly & I'm currently using it in my dev. stack.

Cheers,
Quentin H.

qharnay committed:
Support for custom headers gievn to the browser's requests

clue reviewed (commented):
@qharnay Thanks for filing this PR, this is an interesting one! :+1: 

I'm curious, what is your use case for providing custom HTTP request headers? It's my understanding this isn't possible in the browser-based API, so I wonder how much value it could provide in this implementation.

From a technical perspective, this mostly LGTM, perhaps we could get rid of the `$defaultHeaders` attribute and merge it with the static defaults in `$headers` instead. We should also add some additional tests for overwriting existing headers (`Accept`, `Authorization` etc.)

What do you think? :+1:

qharnay commented:
@clue , thanks for reviewing this one. 

> what is your use case for providing custom HTTP request headers:
I'm using mercure ( https://mercure.rocks ) for a SSE service, on a multi domain app, with private updates, requiring auth. As a multi-domain app, I could not use cookie-based auth and have to use custom HTTP header '"authorization" 

> It's my understanding this isn't possible in the browser-based API
This is totally possible with event-source polyfill : https://www.npmjs.com/package/event-source-polyfill
Which is really widespread on SSE community to ensure browser compliance.

> perhaps we could get rid of the $defaultHeaders attribute and merge it with the static defaults in $headers instead
This property was set to ease tests. but this is not a strong requirement 

> We should also add some additional tests for overwriting existing headers (Accept, Authorization etc.)
This could be a missing case, you're totally right.


I'll  push a new commit ensuring default headers override works.

Thank you @clue !

qharnay committed:
Ensuring default headers can be overriden

qharnay committed:
Inverted test. defualt headers can not be overriden

qharnay commented:
I pushed a tiny test change, ensuring default headers *are not overriden* . In fact the default headers are standard for the SSE and should not be overridden.

Is it enough for you to ensure custom HTTP header ability ?  

cheers !

qharnay commented:
@clue , the CI seems quite broken ...  ( PHP 5.5 error has nothing to do with this PR )

clue commented:
> @clue , the CI seems quite broken ... ( PHP 5.5 error has nothing to do with this PR )

Restart and fixed the temporary hiccup :+1: 

> I pushed a tiny test change, ensuring default headers are not overriden . In fact the default headers are standard for the SSE and should not be overridden.
>
> Is it enough for you to ensure custom HTTP header ability ?

I agree with your reasoning for needing custom HTTP header support. Let's work out some nasty details: What is supposed to happen if I try to overwrite the `Accept` header, what happens if I use alternative casing like `aCCept`? IMO what's most important is that this should be *consistent*, i.e. either accepting or rejecting all forms.

Perhaps we can take a look at how other implementations handle this. What do you think? :+1:

qharnay commented:
Let's work out some nasty details: What is supposed to happen if I try to overwrite the Accept header, what happens if I use alternative casing like aCCept? IMO what's most important is that this should be consistent, i.e. either accepting or rejecting all forms.

qharnay committed:
Custom and default headers are now handled in a mergeHeaders function.
As HTTP headers are case insensitive, this function ensures default headers are not overriden, even by another case

qharnay commented:
I finally came up with the (I think) final version for adding custom HTTP headers / preserving default ones / testing the whole process. Tell me what you think about it @clue. 

[If you're okay, I'll come up with another PR for initializing the Last-Event-Id afterthat ;-) ]

bpolaszek commented:
Dunno how I can add commits to a PR which isn't mine (I probably can't), but here's the rebased version: https://github.com/bpolaszek/reactphp-eventsource/tree/customHeaders

Should I make another PR?

ellisonpatterson commented:
> Here's what this could potentially look like:
> 
> ```
> $browser = new React\Http\Browser();
> $browser = $browser->withProtocolVersion('1.1'); // already supported
> $browser = $browser->withHeader('User-Agent', 'ACME'); // could be useful?
> 
> $es = new Clue\React\EventSource\EventSource('https://example.com/stream', $browser);
> ```
> 
> This isn't currently supported, but seems reasonable and would allow far more packages to take advantage of custom HTTP request headers. What do you think about this?

Would that potentially allow for setting headers that are persistent across all requests made with that browser instance?

SimonFrings commented:
@ellisonpatterson Yes!

clue commented:
I've just filed https://github.com/reactphp/http/issues/449 as a way to keep track of this ticket upstream and believe this is the best way to address this :+1:

In the meantime, I'll assume this one is resolved and will close this for now, please feel free to report back otherwise :+1:
